File: 프로그래머스/연습문제/Lv2/리코쳇 로봇/리코쳇 로봇.cpp
#include <string>
#include <vector>
#include <queue>
using namespace std;

int BFS(vector<string> board, pair<int,int> start){
    vector<vector<int>> visit;
    for(int i=0; i<board.size(); i++){
        vector<int> v(board[0].length(), -1);
        visit.push_back(v);
    }
    
    queue<pair<int,int>> q;
    q.push(start);
    visit[start.first][start.second] = 0;
    
    while(q.size()){
        pair<int,int> tmp = q.front();
        q.pop();
        
        if(board[tmp.first][tmp.second] == 'G')
            return visit[tmp.first][tmp.second];
        
        // 상
        int move_i = tmp.first;
        for(; move_i>=0 && board[move_i][tmp.second]!='D'; move_i--);
        move_i++;
        
        if(visit[move_i][tmp.second] == -1){
            q.push(make_pair(move_i, tmp.second));
            visit[move_i][tmp.second] = visit[tmp.first][tmp.second]+1;
        }
        
        // 하
        move_i = tmp.first;
        for(; move_i<board.size() && board[move_i][tmp.second]!='D'; move_i++);
        move_i--;
        
        if(visit[move_i][tmp.second] == -1){
            q.push(make_pair(move_i, tmp.second));
            visit[move_i][tmp.second] = visit[tmp.first][tmp.second]+1;
        }
        
        // 좌
        int move_j = tmp.second;
        for(; move_j>=0 && board[tmp.first][move_j]!='D'; move_j--);
        move_j++;
        
        if(visit[tmp.first][move_j] == -1){
            q.push(make_pair(tmp.first, move_j));
            visit[tmp.first][move_j] = visit[tmp.first][tmp.second]+1;
        }
        
        // 와
        move_j = tmp.second;
        for(; move_j<board[0].length() && board[tmp.first][move_j]!='D'; move_j++);
        move_j--;
        
        if(visit[tmp.first][move_j] == -1){
            q.push(make_pair(tmp.first, move_j));
            visit[tmp.first][move_j] = visit[tmp.first][tmp.second]+1;
        }
    }
    
    return -1;
}

int solution(vector<string> board) {
    pair<int,int> start;
    for(int i=0; i<board.size(); i++){
        for(int j=0; j<board[0].length(); j++){
            if(board[i][j] == 'R'){
                start.first = i;
                start.second = j;
                break;
            }
        }
    }
    
    return BFS(board, start);
}

File: 프로그래머스/연습문제/Lv2/리코쳇 로봇/리코쳇 로봇.py
def BFS(board, start):
    visit = [[-1 for i in range(len(board[0]))] for i in range(len(board))]
    q = [[start[0], start[1]]]
    visit[start[0]][start[1]] = 0

    while len(q) > 0:
        tmp = q[0]
        q.pop(0)

        if board[tmp[0]][tmp[1]] == 'G':
            return visit[tmp[0]][tmp[1]]
        
        dist = visit[tmp[0]][tmp[1]]
        
        # 상, 좌
        move = [tmp[0], tmp[1]]
        while move[0]>=0 and board[move[0]][tmp[1]]!='D': move[0] -= 1
        while move[1]>=0 and board[tmp[0]][move[1]]!='D': move[1] -= 1
        move[0],move[1] = move[0]+1,move[1]+1

        if visit[move[0]][tmp[1]] == -1:
            q.append([move[0], tmp[1]])
            visit[move[0]][tmp[1]] = dist+1

        if visit[tmp[0]][move[1]] == -1:
            q.append([tmp[0], move[1]])
            visit[tmp[0]][move[1]] = dist+1

        # 하, 우
        move = [tmp[0], tmp[1]]
        while move[0]<len(board) and board[move[0]][tmp[1]]!='D': move[0] += 1
        while move[1]<len(board[0]) and board[tmp[0]][move[1]]!='D': move[1] += 1
        move[0],move[1] = move[0]-1,move[1]-1

        if visit[move[0]][tmp[1]] == -1:
            q.append([move[0], tmp[1]])
            visit[move[0]][tmp[1]] = dist+1

        if visit[tmp[0]][move[1]] == -1:
            q.append([tmp[0], move[1]])
            visit[tmp[0]][move[1]] = dist+1

    return -1

def solution(board):
    start = []
    for i in range(len(board)):
        if 'R' in board[i]:
            start.append(i)
            start.append(board[i].index('R'))
            break

    return BFS(board, start)
